rllib/execution/parallel_requests.py

```diff
@@ -178,7 +180,237 @@ def wait_asynchronous_requests(
+            elif len(self._all_workers) == 1:
+                actor = list(self._all_workers)[0]
+            else:
+                # load balance by getting 2 random actors and then placing the call
+                # on the actor with the least number of inflight requests. If max
+                # inflight requests is reached, place the call on the actor with the
+                # smaller queue
+                [a1, a2] = random.sample(self._all_workers, 2)
+                num_inflight_req_a1 = len(self._remote_requests_in_flight[a1])
+                num_inflight_req_a2 = len(self._remote_requests_in_flight[a2])
+                if num_inflight_req_a1 < num_inflight_req_a2:
+                    actor = a1
+                elif num_inflight_req_a1 > num_inflight_req_a2:
+                    actor = a2
+                else:
+                    if self._call_queue[a1].qsize() < self._call_queue[a2].qsize():
+                        actor = a1
+                    else:
+                        actor = a2
+            self._call_queue[actor].put(call)
+
+        self._run()
+
+    def _run(self) -> int:
+        """Launch the submited requests remotely on workers that don't have more
+        than max_inflight_requests in flight.
+        """
+        num_requests_launched = 0
+        for actor, q in self._call_queue.items():
+            # print(len(self._remote_requests_in_flight[actor]))
+            # print(self._max_remote_requests_in_flight)
+            while (
+                len(self._remote_requests_in_flight[actor])
+                < self._max_remote_requests_in_flight
+            ) and q.qsize() > 0:
+                remote_fn, fn_args, fn_kwargs = q.get()
+                req = actor.apply.remote(remote_fn, *fn_args, **fn_kwargs)
+                self._remote_requests_in_flight[actor].add(req)
+                self._pending_to_actor[req] = actor
+                self._pending_remotes.append(req)
+                num_requests_launched += 1
+        return num_requests_launched
+
+    def get_ready_results(self) -> Dict[ActorHandle, List[Any]]:
+        """Get results that are ready to be returned
+
+        Returns:
+            A dictionary of actor handles to lists of returns from tasks that were
+             previously submitted to this actor pool that are now ready to be returned.
+             If return_object_refs
+
+        """
+        ready_requests_dict = defaultdict(list)
+        ready_requests, self._pending_remotes = ray.wait(
+            self._pending_remotes,
+            timeout=0,
+            num_returns=len(self._pending_remotes),
+        )
+        if not self._return_object_refs:
+            objs = ray.get(ready_requests)
+        else:
+            objs = ready_requests
+        for req, obj in zip(ready_requests, objs):
+            actor = self._pending_to_actor[req]
+            self._remote_requests_in_flight[actor].remove(req)
```

Comment:
do we need to mutex lock any of the state variables ... ?

Reply:
offline conversation: 

probably not, since python objects are threadsafe anyways.